### ISSUE TITLE
oidentd returns user "root" for all requests on FreeBSD 13 

### ISSUE THREAD

**Issue by aalewis** [janikrabe/oidentd#37]
Per thread title, after updating to FreeBSD 13, oidentd no longer differentiates connections and returns the owner user as root on all ident requests.

oident installed using the standard "pkg install oidentd" functionality, was working as expected on 12.x.

Would be happy to help debug.

**Comment by janikrabe** [comment]
Hi @aalewis!

Could you try running oidentd interactively with the `-id` flags? The `-d` will cause more debugging information to be printed. Support for this flag currently still needs to be enabled at compile time (with `./configure --enable-debug`) so if FreeBSD's package was compiled without debugging support you may have to build oidentd yourself. The `INSTALL` file contains instructions on how to do that, but the short version is `./autogen.sh && ./configure --enable-debug && make && src/oidentd -id`.

With debugging enabled, oidentd should print something like this when it receives an Ident query:

```
Connection from localhost (::1):55555
[localhost] Successful lookup: 55555 , 113 : realuser (root)
```

Depending on whether `realuser` is also `root` this is either a configuration issue or a bug in the kernel module.

**Comment by aalewis** [comment]
```
# oidentd -id

Warning: privileges will not be dropped automatically because oidentd needs to run as root on this system
[inet_util.c:75:setup_bind] DEBUG: address family 28 not supported
Connection from orwell.freenode.net (185.30.166.37):33151
[orwell.freenode.net] Successful lookup: 46447 , 6697 : root (UNKNOWN)
```

**Comment by janikrabe** [comment]
Thank you for reporting this! I'm seeing something similar (although not exactly the same) on FreeBSD 13. I'll have a look this weekend.

```
[inet_util.c:75:setup_bind] DEBUG: address family 28 not supported
```

This is probably unrelated, but shouldn't be happening, and I'm not able to reproduce this. Could you share the output of `src/oidentd --version` please?

If possible, could you also share your configuration file? (This is likely in `/usr/local/etc/` if you installed from source; the path should be shown in the `--version` output).

**Comment by aalewis** [comment]
Sorry, I think I conflated two issues, potentially one in the /usr/ports/security/oidentd system. Building a modified /usr/ports/security/oidentd with debug added and run:
```
$ ./oidentd --version
oidentd 2.5.0 by Janik Rabe <oidentd@janikrabe.com>
Originally written by Ryan McCabe <ryan@numb.org>
https://oidentd.janikrabe.com

Build information:
        Kernel driver:         freebsd4.c
        Needs kmem access:     No
        Needs root access:     Yes
        Debug build:           Yes
        Masquerading support:  No
        IPv6 support:          No
        Linux libnfct support: No
        UDB library support:   No

Build settings:
        Configuration directory:   /usr/local/etc
        User configuration file:   ~/.config/oidentd.conf
        User configuration file:   ~/.oidentd.conf

# ./oidentd -id
Warning: privileges will not be dropped automatically because oidentd needs to run as root on this system
[inet_util.c:75:setup_bind] DEBUG: address family 28 not supported
Connection from rothfuss.freenode.net (84.17.44.180):57351
[rothfuss.freenode.net] Successful lookup: 20070 , 6697 : root (UNKNOWN)
```
Building and running straight from source:
```
$ ./oidentd --version
oidentd 2.5.0 by Janik Rabe <info@janikrabe.com>
Originally written by Ryan McCabe <ryan@numb.org>
https://janikrabe.com/projects/oidentd/

Build information:
        Kernel driver:         freebsd4.c
        Needs kmem access:     No
        Needs root access:     Yes
        Debug build:           Yes
        Masquerading support:  No
        IPv6 support:          Yes
        Linux libnfct support: No

Build settings:
        Configuration directory:   /usr/local/etc
        User configuration file:   ~/.config/oidentd.conf
        User configuration file:   ~/.oidentd.conf

# ./oidentd -id
Warning: privileges will not be dropped automatically because oidentd needs to run as root on this system
Connection from verne.freenode.net (185.30.166.38):56291
[oidentd.c:351:service_request] DEBUG: getpwuid(23675): No error: 0
```
Clearly there's some kind of discrepancy in the FreeBSD ports build system/patches. TL;DR:

```
FreeBSD ports oidentd -> works, but throws an address family error and only returns root
Github source oidentd -> doesn't seem to work
```

**Comment by aalewis** [comment]
Config:

```
default {
        default {
                allow spoof
                allow spoof_all
                deny spoof_privport
                deny random_numeric
                deny numeric
                deny hide
        }
}

user root {
        default {
                force reply "UNKNOWN"
        }
}
```

**Comment by janikrabe** [comment]
Thank you! Could you try building from the `freebsd-xucred` branch please? That should fix it. We were using FreeBSD's kernel-internal `ucred` structure instead of the external `xucred`. This happened to work because the first couple of bytes were identical, but that is no longer the case in FreeBSD 13.

The message about address family 28 being unsupported is because oidentd was built without IPv6 support. This is safe to ignore if you don't need to look up IPv6 connections. If you do, a `make config` (and reinstall) in the port skeleton should [allow you to enable IPv6](https://svnweb.freebsd.org/ports/head/security/oidentd/Makefile?revision=543953&view=markup#l25).

**Comment by aalewis** [comment]
On the branch: still no ident response and getting
`[oidentd.c:351:service_request] DEBUG: getpwuid(23675): No error: 0`

**Comment by janikrabe** [comment]
Sorry, I forgot to mention that changes to `src/kernel/*` aren't recompiled automatically when you run `make`. The kernel module is copied to `src/os.c` by the `./configure` script.

Something like this should work (at least it does for me):

```
git fetch
git checkout freebsd-xucred
git reset --hard origin/freebsd-xucred
./configure
make
src/oidentd -Sid
```

**Comment by aalewis** [comment]
Confirmed it looks like that resolves it.

**Comment by janikrabe** [comment]
Great, merging the change then. Thank you for reporting and helping debug this!

**Comment by aalewis** [comment]
Thank you for investigating and being so responsive! Much appreciated.

**Comment by janikrabe** [comment]
I've backported this to v2.5.x and released v2.5.1 which includes the fix.

**Comment by mateuszkwiatkowski** [comment]
Bug report on updating a port: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=256230
